### ISSUE TITLE
[lecture12][0521] About ReLU6, ShuffleNet, and More

### ISSUE THREAD

**Issue by Burugi** [PiLab-CAU/ComputerVision-2401#30]
Hello, professor, I'm raising this issue because I'd like to review what we covered in class today about ReLU6, ShuffleNet, etc. and make sure I'm correct. I'd also like to ask if you know anything about the Shift method for computational reduction.

This issue is a complication of other questions, so I hope that this doesn't confuse anyone, as only the second question is included in the class, and the third question is for private study.

Here's the question in a nutshell (the questions are labeled with numbers)

"ReLU"
![image](https://github.com/PiLab-CAU/ComputerVision-2401/assets/52272632/0d296ea1-cdb5-4a11-801f-2b3a855f9cb2)

ReLU6 is min(max(0, x), 6), where the value above a certain point is 6.
1. I was told that the value of 6 is an experimental result and not because 6 is a special number, but I wonder if this is correct. Also, when I searched for it, I found that it is used as 8 in certain experiments.([link1](https://stackoverflow.com/questions/47220595/why-the-6-in-relu6))

"ShuffleNet"

![image](https://github.com/PiLab-CAU/ComputerVision-2401/assets/52272632/12d39999-fcb7-4878-a26e-ccc3552b8c68)

2. I understood that ShuffleNet is a groupwise convolution that I learned before, and after 1x1 among the groups, the feature maps that came out are distributed one per group to create new groups, and this is the convolution, but I wonder if it is correct.

-----------------------------------------------------------------------------------------------------------------------

"Shift"

![image](https://github.com/PiLab-CAU/ComputerVision-2401/assets/52272632/0e3e4a47-eb88-4cc2-a401-73c1ca4be6ed)

This is something I came across while looking for methods for computational reduction, and the original paper referenced the following. (Wu, Bichen, et al. "Shift: A zero flop, zero parameter alternative to spatial convolutions." Proceedings of the IEEE conference on computer vision and pattern recognition. 2018.)

In a nutshell, the convolution is performed by shifting the feature map that comes out of the convolution up, down, left, and right, and excluding the parts that are beyond the original range.

3. I'm wondering if this method has been used in practice, and what are some other methods for computational reduction other than shufflenet and shift that you've used or know of?

Thank you.

Hyunwoong LIM

**Comment by yjyoo3312** [comment]
@Burugi 

Thank you for the comment!

1. ReLU6: As far as I know, the number 6 was set experimentally. While different values might affect accuracy, we believed the difference would not be significant, so we did not test other settings during my time at the company.

2. "Shufflenet: The basic structure of Shufflenet you understand is correct. However, another important aspect of Shufflenet is the **channel-shuffle operation**. This operation shuffles the channels to distribute information across each convolution group. Despite its effectiveness, the channel-shuffle operation is challenging to optimize, making it less widely used in the field compared to the Mobilenet family.

3. ShiftNet: There have been many attempts to replace or complement the spatial operation part of the convolution (as many researchers have questioned the actual role of depthwise convolution). Some notable attempts include:

- Shift operation
[Shift: A Zero FLOP, Zero Parameter Alternative to Spatial Convolutions](https://arxiv.org/abs/1711.08141)
[All You Need is a Few Shifts: Designing Efficient Convolutional Neural Networks for Image Classification](https://arxiv.org/abs/1903.05285)
[Addressnet: Shift-based primitives for efficient convolutional neural networks
](https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=8658953)
- Linear transformation 
[GhostNet: More Features from Cheap Operations](https://arxiv.org/abs/1911.11907)
- maxpool
[Learning features with parameter-free layers](https://arxiv.org/abs/2202.02777)

However, all attempts to apply computationally cheaper operations have not been widely adopted in practice due to several reasons:

- These operations are not sufficiently optimized in CUDA or other low-level implementations, so they are not significantly faster, and can even be slower than traditional convolution or depthwise convolution operations.
- When serving the model, we usually convert our PyTorch model to more optimized formats like ONNX to TensorRT, OpenVino, TFLite, NCNN, or others. However, these conversions do not work well with less common operations beyond standard convolution.

Despite this, I believe these studies are meaningful as they enhance our understanding of the role of spatial and depthwise convolution in lightweight convolutional networks.
